omnichain/multioutput/contracts/MultiOutput.sol

```diff
@@ -74,21 +70,105 @@ contract MultiOutput is zContract, Ownable {
+                if (btcRecipient.length == 0) revert InvalidRecipient();
+                recipient = abi.encodePacked(btcRecipient);
+            }
+
+            _doSwapAndWithdraw(zrc20, amountToTransfer, targetZRC20, recipient);
+        }
+    }
+
+    function _doSwapAndWithdraw(
+        address zrc20,
+        uint256 amountToTransfer,
+        address targetZRC20,
+        bytes memory recipient
+    ) internal {
+        uint256 outputAmount = SwapHelperLib._doSwap(
+            systemContract.wZetaContractAddress(),
+            systemContract.uniswapv2FactoryAddress(),
+            systemContract.uniswapv2Router02Address(),
+            zrc20,
+            amountToTransfer,
+            targetZRC20,
+            0
+        );
+        _doWithdrawal(
+            targetZRC20,
+            outputAmount,
+            recipient
+        );
+        emit Withdrawal(targetZRC20, outputAmount, recipient);
+    }
+
+    function parseMessage(
+        uint256 chainID,
+        bytes calldata message)
+        public
+        pure
+        returns (address, bytes memory, address[] memory)
+    {
+        address evmRecipient;
+        bytes memory btcRecipient;
+        address[] memory destinationTokens;
+        if (chainID == BITCOIN) {
+            evmRecipient = BytesHelperLib.bytesToAddress(message, 0);
+            uint256 numTokens = message.length / 20 - 1;
+            destinationTokens = new address[](numTokens);
+            for (uint256 i = 0; i < numTokens; i++) {
+                destinationTokens[i] = BytesHelperLib.bytesToAddress(
+                    message,
+                    20 + i * 20
+                );
+            }
+        } else {
+            (address evmAddress, bytes memory btcAddress, bytes memory targetTokens) = abi.decode(
+                message,
+                (address, bytes, bytes)
             );
-            emit Withdrawal(targetZRC20, outputAmount, recipient);
+
+            btcRecipient = btcAddress;
+            evmRecipient = evmAddress;
+
+            uint256 numTokens = targetTokens.length / 32; 
+            destinationTokens = new address[](numTokens);
+            for (uint256 i = 0; i < numTokens; i++) {
+                destinationTokens[i] = _bytesMemoryToAddress(targetTokens, i * 32);
+            }
+        }
+
+        return (evmRecipient, btcRecipient, destinationTokens);
+    }
+
+        function _doWithdrawal(
+        address targetZRC20,
+        uint256 amount,
+        bytes memory receipient
+    ) internal {
+        (address gasZRC20, uint256 gasFee) = IZRC20(targetZRC20)
+            .withdrawGasFee();
+
+        if (gasZRC20 != targetZRC20) revert WrongGasContract();
+        if (gasFee >= amount) revert NotEnoughToPayGasFee();
+
+        IZRC20(targetZRC20).approve(targetZRC20, gasFee);
+        IZRC20(targetZRC20).withdraw(
+            abi.encodePacked(receipient),
+            amount - gasFee
+        );
+    }
+
+    function _bytesMemoryToAddress(
```

Comment:
Can we reuse an existing function from the toolkit?

https://github.com/zeta-chain/toolkit/blob/953c05ed834d3481acd63d4ea1c9faec75168ab2/contracts/BytesHelperLib.sol#L15-L26

Reply:
I have used it before, but found that it does not work, for test:
https://github.com/zeta-chain/example-contracts/blob/ea886e07ed619d19c8d4dd2f0d922bea1dc7ba64/omnichain/multioutput/test/MultiOutput.t.sol#L17

Reply:
Oh, good to know. We can keep it in this PR. I've submitted a PR to the toolkit to add this function there, and when we release a new version we can import this function (I'm just trying to keep tutorials as lean as possible).

https://github.com/zeta-chain/toolkit/pull/90